### DISCUSSION TITLE
Missing SMBUS controllers

### DISCUSSION THREAD

**Discussion by x86fr** [memtest86plus/memtest86plus#86] [Show and tell]
Following discussions, I've just run some stats on the CPU-Z database to grab all SMBUS Controllers and compare them with what we currently support. As the amount of data is enormous, I only parsed March 2022 (~500K entries). 

Here are the results:

VID | DID | Count | % | Name | Supported |  Link
-- | -- | -- | -- | -- | -- | --
8086 | 7AA3 | 35502 | 8.32% | Alder Lake-S (PCH) | Yes |  
8086 | 43A3 | 28028 | 6.57% | Tiger Lake-H (PCH) | Yes |  
8086 | A323 | 23697 | 5.55% | Cannon Lake-H (PCH) | Yes |  
8086 | 8C22 | 20327 | 4.76% | Lynx Point (PCH) | Yes |  
8086 | 1C22 | 20016 | 4.69% | 6 Series (PCH) | Yes |  
8086 | A2A3 | 19258 | 4.51% | Kaby Lake (PCH-H) | Yes |  
8086 | A123 | 17128 | 4.01% | Sunrise Point-H (PCH) | Yes |  
8086 | 1E22 | 17073 | 4.00% | Panther Point (PCH) | Yes |  
8086 | 06A3 | 15097 | 3.54% | Comet Lake-H (PCH) | Yes |  
8086 | 9D23 | 12100 | 2.84% | Sunrise Point-LP   (PCH) | Yes |  
8086 | 3B30 | 6935 | 1.63% | 5/3400 Series (PCH) | Yes |  
8086 | 27DA | 6725 | 1.58% | 82801G (ICH7) | Yes |  
8086 | A3A3 | 6280 | 1.47% | Comet Lake-V (PCH) | Yes |  
8086 | A0A3 | 6080 | 1.42% | Tiger Lake-LP (PCH) | Yes |  
8086 | 3A30 | 5344 | 1.25% | ICH10 | Yes |  
8086 | 8D22 | 4717 | 1.11% | Wellsburg (PCH) | Yes |  
8086 | 2930 | 4130 | 0.97% | 82801I (ICH9) | Yes |  
8086 | 1D22 | 4009 | 0.94% | Patsburg (PCH) | Yes |  
8086 | 8CA2 | 3961 | 0.93% | Wildcat Point (PCH) | Yes |  
8086 | 9C22 | 3429 | 0.80% | Lynx Point-LP (PCH) | Yes |  
8086 | 9CA2 | 3156 | 0.74% | Wildcat Point-LP   (PCH) | Yes |  
8086 | 9DA3 | 2700 | 0.63% | Cannon Lake-LP (PCH) | Yes |  
8086 | 02A3 | 2487 | 0.58% | Comet Lake (PCH) | Yes |  
8086 | 34A3 | 2408 | 0.56% | Ice Lake-LP (PCH) | Yes |  
8086 | 31D4 | 1610 | 0.38% | Gemini Lake (SOC) | Yes |  
8086 | 283E | 1240 | 0.29% | 82801H (ICH8) | Yes |  
8086 | 0F12 | 1069 | 0.25% | BayTrail (SOC) | Yes |  
8086 | 2292 | 959 | 0.22% | Braswell  (SoC) | Yes | 
8086 | 51A3 | 839 | 0.20% | Alder Lake-P (PCH) | Yes |  
8086 | 3A60 | 812 | 0.19% | ICH10 | Yes |  
8086 | 5AD4 | 809 | 0.19% | Apollo Lake (SoC) | **NO** | https://valid.x86.fr/zzrirl
8086 | 4DA3 | 405 | 0.09% | Jasper Lake (SOC) | Yes |  
8086 | A1A3 | 231 | 0.05% | Lewisburg (PCH) | Yes |  
8086 | 269B | 157 | 0.04% | 6310ESB/6320ESB | Yes |  
8086 | 266A | 84 | 0.02% | 82801F (ICH6) | Yes |  
8086 | 24D3 | 34 | 0.01% | 82801E (ICH5) | Yes |  
8086 | 24C3 | 18 | 0.00% | 82801DB (ICH4) | Yes |  
8086 | 2443 | 15 | 0.00% | 82801BA (ICH2) | Yes |  
8086 | 2483 | 15 | 0.00% | 82801CA (ICH3) | Yes |  
8086 | 4B23 | 13 | 0.00% | Elkhart Lake (PCH) | Yes |  
8086 | 38A3 | 4 | 0.00% | Ice Lake-N (PCH) | Yes |  
8086 | 2413 | 3 | 0.00% | 82801AA (ICH) | Yes |  
8086 | 19DF | 2 | 0.00% | Denverton  (SOC) | Yes |  
8086 | 1F3C | 1 | 0.00% | Avoton (SOC) | Yes |  
8086 | 25A4 | 1 | 0.00% | 6300ESB | Yes |  
1022 | 790B | 125365 | 29.38% | AMD FCH (Zen) | Yes |  
1022 | 780B | 7077 | 1.66% | AMD FCH (Pre-Zen) | Yes |  
1022 | 746A | 29 | 0.01% | AMD 8111 | **NO** | https://valid.x86.fr/r8n8y0
1002 | 4385 | 12550 | 2.94% | ATI SB600+ (Now AMD) | Yes |  
1002 | 4372 | 85 | 0.02% | ATI SB400 | **NO** | https://valid.x86.fr/p0pevj
1002 | 4353 | 3 | 0.00% | ATI SB200 | **NO** | https://valid.x86.fr/zq1s8m
10DE | 03EB | 1746 | 0.41% | NVIDIA nForce 630a | **NO** | https://valid.x86.fr/zhdh8p
10DE | 0752 | 277 | 0.06% | NVIDIA nForce 720a | **NO** | https://valid.x86.fr/ml8yrj
10DE | 0368 | 133 | 0.03% | NVIDIA nForce 790i   Ultra | **NO** | https://valid.x86.fr/sucjma
10DE | 0264 | 127 | 0.03% | NVIDIA nForce 410/430   MCP | **NO** | https://valid.x86.fr/x3ar9p
10DE | 0AA2 | 119 | 0.03% | NVIDIA nForce 730i | **NO** | https://valid.x86.fr/sj5idk
10DE | 07D8 | 75 | 0.02% | NVIDIA nForce 630i | **NO** | https://valid.x86.fr/kxk2q7
10DE | 0052 | 73 | 0.02% | NVIDIA nForce4 MCP | **NO** | https://valid.x86.fr/y8es4r
10DE | 0446 | 56 | 0.01% | NVIDIA nForce 520 | **NO** | https://valid.x86.fr/tgqtkk
10DE | 0542 | 42 | 0.01% | NVIDIA nForce 560 | **NO** | https://valid.x86.fr/t25i13
10DE | 0064 | 28 | 0.01% | NVIDIA nForce2 MCP | **NO** | https://valid.x86.fr/zvc56u
10DE | 0D79 | 9 | 0.00% | NVIDIA GeForce 320M | **NO** | https://valid.x86.fr/5iuxhj
10DE | 00E4 | 6 | 0.00% | NVIDIA nForce3 MCP | **NO** | https://valid.x86.fr/q1bd6q
10DE | 0084 | 1 | 0.00% | NVIDIA nForce2 MCP   (Mobile) | **NO** | https://valid.x86.fr/i5jyay
10DE | 01B4 | 1 | 0.00% | NVIDIA nForce  MCP | **NO** | https://valid.x86.fr/s59nzv
1D94 | 790B | 6 | 0.00% | AMD FCH (Zen) | Yes |  
1039 | 0016 | 1 | 0.00% | SiS 963 | **NO** |  

We have SMBUS support for 99.15% of the total. Quite nice! Of course, some platforms still requires proprietary tricks to work, but that's not bad

I will now focus on the 3 most common chipset that we don't support yet: 

- nVidia nForce 630a
- Intel Apollo Lake. 
- nVidia nForce 720a

**Comment by debrouxl** [comment]
* the Linux i2c-nforce2.c driver suggests that the NVIDIA SMBus controllers are pretty homogeneous, with a single test for specific device IDs, so adding support for the still reasonably popular 10DE:03EB MCP61 ought to effectively add support for all of the others of the same class;
* the CPU temperature reported in https://valid.x86.fr/zhdh8p is... interesting :) The figure is less unrealistic as a °K temperature, but still overly low, unless the user is using cryogenic cooling, which is a bit excessive for a 95W TDP CPU;
* I see 8086:2292 in the current list (HEAD @ 67cb78f4e5c9cf2fdab195d65e332d965a41c5eb );
* the 8086:5AD4 Apollo Lake is listed in Linux i2c-i801.c and my update to https://github.com/anphsw/memtest86/blob/master/spd.c , but not in current smbus.c, indeed.

**Reply by x86fr** [reply]
- Hope so. I've just ordered an Asus M4N68T-MLE V2 on eBay to check that. I have nForce2/3 available here, and almost some of the infamous Intel chipset from Nvidia.
- The actual core's temperature is 43°C but for some reason, it also tries to read a non-existent temperature sensor on Bulldozer and display this one. Fix planned.
- Right. Corrected. Now > 99% supported 😅
- IIRC, Apollo Lake uses a different protocol. The Linux drivers are misleading because those are the drivers to access the SMBus controller on the PCH. On many CPUs, like Intel's -E and -X, the SPD is not located on the PCH but directly on a side SMBUS Controller on the CPU.

**Comment by debrouxl** [comment]
* the GA-M61P-S3 I recently bought also has 10de:03eb ;
* 1002:4353, 1002:4363 and 1002:4372 are handled by Linux i2c-piix4.c.

**Comment by x86fr** [comment]
Any test welcome :)

**Comment by debrouxl** [comment]
Hopefully tonight on the GA-M61P-S3, R815 & RS904A.

**Reply by x86fr** [reply]
Cool. Don't expect much on servers as the SPD SMBUS is probably on the PCH

**Reply by oed-streicherd** [reply]
Ist there a easy why to talk SMBus on Intel Server with current Xeon Procs?

**Reply by x86fr** [reply]
Intel -E and -X CPUs (starting with SNB-E) require a specific code to access SPD from a proprietary smbus controller on the CPU.

**Comment by 01e3** [comment]
I just tested memtest on Dell Optiplex 740 (nForce 430 MCP) and SPD detection did not work.

`00:0a.1 SMBus [0c05]: NVIDIA Corporation MCP51 SMBus [10de:0264] (rev a3)`

SMBus and SPD works on Linux:

```
# i2cdetect -y 13
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         08 -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: 50 51 52 53 -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --

# i2cdump -y 13 0x50
No size specified (using byte-data access)
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f    0123456789abcdef
00: 80 08 08 0e 0a 61 40 00 05 25 40 00 82 08 00 00    ?????a@.?%@.??..
10: 0c 08 38 01 02 00 03 3d 50 50 60 32 1e 32 2d 01    ??8??.?=PP`2?2-?
20: 17 25 05 12 3c 1e 1e 00 36 39 7f 80 14 1e 00 00    ?%??<??.69????..
30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 12 f9    ..............??
40: 7f 98 00 00 00 00 00 00 06 32 47 2d 55 44 49 4d    ??......?2G-UDIM
50: 4d 00 00 00 00 00 00 00 00 00 00 00 00 09 09 75    M............??u
60: cc c2 b0 00 00 00 00 00 00 00 00 00 00 00 00 00    ???.............
70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
f0: 39 39 30 35 34 32 39 2d 30 30 37 2e 41 30 30 4c    9905429-007.A00L
```

Should I just try uncommenting 0x0264?

```
        case VID_NVIDIA:
            switch(did)
            {
                // case 0x01B4: // nForce
                case 0x0064:    // nForce 2
                // case 0x0084: // nForce 2 Mobile
                // case 0x00E4: // nForce 3
                // case 0x0034: // MCP04
                // case 0x0052: // nForce 4
                // case 0x0264: // nForce 410/430 MCP
                case 0x03EB:    // nForce 630a
                // case 0x0446: // nForce 520
                // case 0x0542: // nForce 560
                case 0x0752:    // nForce 720a
                // case 0x07D8: // nForce 630i
                // case 0x0AA2: // nForce 730i
                // case 0x0D79: // MCP89
                // case 0x0368: // nForce 790i Ultra
                    return nv_mcp_get_smb();
                default:
                    return false;
            }
            break;

```

**Reply by x86fr** [reply]
Yes you can try, but be warned that some serious issues may come. 

Let us know if you try. ;)

**Reply by debrouxl** [reply]
Indeed, there might be issues. However, the nForce2 family is quite homogeneous and devoid of quirks, at least according to the Linux "i2c-nforce2" driver, so chances are that it will work like the three other PCI DIDs currently supported :)

**Reply by 01e3** [reply]
It "just worked".
![Dell-Optiplex-740-spd](https://user-images.githubusercontent.com/105136494/174979117-cc5ab6f8-06d9-40d3-a6e0-ea0fdd3a372f.png)

Do you want a patch or going to update it next time you touch the file?

**Reply by 01e3** [reply]
Also, is there a separate place to track temperature reporting, which is also missing?

**Reply by x86fr** [reply]
I'm working on nForce Intel so I'll uncomment it in the process.
About temperature, what's the CPUID of your CPU? There is many different X2 5000+ with different core.

**Reply by 01e3** [reply]
I actually have two similar systems, but with slightly different CPUs:

https://valid.x86.fr/2yz1kg
https://valid.x86.fr/epmtak

In both cases there is no temp reported on Memtest+ but note that CPU-Z reports the CPU temperature correctly, same for Linux.

**Comment by debrouxl** [comment]
The SMBus controllers for multiple nForce models are now supported, as are several variants of PIIX4, some of which did not appear in the above list (multiple VIA PCI IDs for viapro controllers, ALi M1563), and some older non-PIIX4-like ALi controllers as well :)

Maybe 1039:0016 some day ? It appeared in the above list, albeit with a single occurrence. My ECS K7S5A (I read that it was a popular model in its time) has this SiS 96x SMBus controller.

**Reply by x86fr** [reply]
I have kept SiS development boards with 963/964 Southbridge, so yup 👍 But IMC registers read and ECC support is next on my list right now.
